esmlab/statistics.py

```diff
@@ -18,40 +18,56 @@
 )
 
 
-def _apply_nan_mask(x, weights, avg_over_dims_v):
-    weights = weights.where(x.notnull())
-    np.testing.assert_allclose(
-        (weights / weights.sum(avg_over_dims_v)).sum(avg_over_dims_v), 1.0
-    )
-    return weights
+def _get_weights_and_dims(x, y=None, weights=None, dim=None):
+    """ Get weights and dimensions """
 
+    if dim and isinstance(dim, str):
+        dims = [dim]
 
-def _apply_nan_mask_for_two_arrays(x, y, weights):
-    valid = x.notnull() & y.notnull()
-    weights = weights.where(valid)
-    return weights
+    elif isinstance(dim, list):
+        dims = dim
+
+    else:
+        dims = [k for k in x.dims]
+
+    op_over_dims = [k for k in dims if k in x.dims]
+    if not op_over_dims:
+        raise ValueError("Unexpected dimensions for variable {0}".format(x.name))
+
+    dims_shape = tuple(l for i, l in enumerate(x.shape) if x.dims[i] in op_over_dims)
+    if weights is None:
```

Comment:
If we are to restore the optional application of `_apply_nan_mask`, we might want to pull it back out into a function:
```python
def _apply_nan_mask(weights, x, y=None):
    if y is not None and isinstance(y, xr.DataArray):
        assert x.shape == y.shape
        valid = x.notnull() & y.notnull()
    else:
        valid = x.notnull()
    return weights.where(valid)
```

Then this "if" statement might look like this:
```python
    if weights is None:
        weights = xr.DataArray(np.ones(dims_shape), dims=op_over_dims)
        weights = _apply_nan_mask(weights, x, y)
     else:
        assert weights.shape == dims_shape
        if apply_nan_mask:
           weights = _apply_nan_mask(weights, x, y)
```

Reply:
This essentially ignores the value of `apply_nan_mask` if weights are not specified. Perhaps there is a better solution?

Reply:
To avoid code duplication: `weights = _apply_nan_mask(weights, x, y)`, I changed the `if` statement into this:

```python
if weights is None:
    weights = xr.DataArray(np.ones(dims_shape), dims=op_over_dims)
else:
    assert weights.shape == dims_shape

if apply_nan_mask:
    weights = _apply_nan_mask(weights, x, y)
```

Reply:
Let me know if this introduces any inconvenience

Reply:
@andersy005: this will likely be problematic. If we are generating the weights and they are all ones, but the field has missing values, our normalized weights will not sum to one unless they are masked. Thus, it the weights are not provided, I think we have to apply the nan mask.

Reply:
This is now fixed and ready for review.